frontend/src/Frontend/VersionedStore/V2.hs

```diff
@@ -0,0 +1,218 @@
+      let regenerated = KeyPair
+            { _keyPair_publicKey = public
+            , _keyPair_privateKey = Just private
+            }
+      pure ([(keyIdx, Key regenerated)], mempty)
+
+    splitOldKey (keyIdx, V0.SomeAccount_Account a) = pure
+      ([(keyIdx, Key (extractKey a))]
+      , oldAccountToNewStorage a
+      )
+
+    oldAccountToNewStorage :: V0.Account key -> AccountStorage
+    oldAccountToNewStorage a =
+      let
+        accountNameText = V0.unAccountName . V0._account_name $ a
+        chainIdText = V0.unChainId . V0._account_chainId $ a
+        newChainId = ChainId chainIdText
+        accountNotesText = V0.unAccountNotes . V0._account_notes $ a
+        newAccountNotes = mkAccountNotes accountNotesText
+        newUnfinishedXChain = V0._account_unfinishedCrossChainTransfer a
+
+        accounts = Map.singleton (AccountName accountNameText) $ AccountInfo Nothing
+          $ Map.singleton newChainId $ VanityAccount newAccountNotes newUnfinishedXChain
+
+      in AccountStorage $ Map.singleton (V0._account_network a) accounts
+
+    upgradePublicKey = PublicKey . V0.unPublicKey
+
+    extractKey (V0.Account { V0._account_key = kp } ) = KeyPair
+      -- This relies on the V0.Wallet.PublicKey FromJSON checking that it is Base16!
+      { _keyPair_publicKey = upgradePublicKey $ V0._keyPair_publicKey kp
+      , _keyPair_privateKey = V0._keyPair_privateKey kp
+      }
+
+upgradeFromV1 :: (Monad m, HasCrypto key m) => DMap (V1.StoreFrontend key) Identity -> m (DMap (StoreFrontend key) Identity)
+upgradeFromV1 v1 =
+    pure $ DMap.fromList  . catMaybes $
+      [
+        copyKeyDSum V1.StoreFrontend_Network_PublicMeta StoreFrontend_Network_PublicMeta v1
+      , copyKeyDSum V1.StoreFrontend_Network_SelectedNetwork StoreFrontend_Network_SelectedNetwork v1
+      , copyKeyDSum V1.StoreFrontend_OAuth_Tokens StoreFrontend_OAuth_Tokens v1
+      , copyKeyDSum (V1.StoreFrontend_OAuth_State OAuthProvider_GitHub) (StoreFrontend_OAuth_State OAuthProvider_GitHub) v1
+      , copyKeyDSum V1.StoreFrontend_Wallet_Keys StoreFrontend_Wallet_Keys v1
+      , copyKeyDSum V1.StoreFrontend_Wallet_Accounts StoreFrontend_Wallet_Accounts v1
+      , copyKeyDSum V1.StoreFrontend_ModuleExplorer_SessionFile StoreFrontend_ModuleExplorer_SessionFile v1
+      , newNetworks
+      ]
+  where
+    newNetworks = DMap.lookup V1.StoreFrontend_Network_Networks v1
+      <&> \nets -> StoreFrontend_Network_Networks :=> convertNodeRefs <$> nets
+
+toMultiSet :: Ord a => [a] -> Map a Int
+toMultiSet = Map.fromList . flip zip (repeat 1)
+
+fromMultiSet :: Ord a => Map a Int -> [a]
+fromMultiSet = ($ []) . Map.foldrWithKey (\k i -> (.) (dlrep k i)) id
+  where
+    dlrep v n
+      | n < 0 = error "fromMultiSet: IMPOSSIBLE"
+      | n == 0 = id
+      | otherwise = (v:) . dlrep v (n - 1)
+
+convertNodeRefs :: Map NetworkName [NodeRef] -> Map NetworkName [NodeRef]
+convertNodeRefs = Map.mapWithKey migrate
+  where
+    migrate = \case
```

Comment:
Hmmmm...I'm contemplating the possibility of doing the replacement no matter what the network name is.  I think I've seen instances where it's lowercase "mainnet" rather than "Mainnet".  We would have to keep track of which of the node refs were actually found to know which new ref to add.  But that would work for these other cases.